### PR TITLE
CT-429 Gov UK search api client and response

### PR DESCRIPTION
**Pull request by stephenrichards**
This PR adds a GOVUK search client and response classes into the
code base, but as yet, nothing is wired up to use them

### CODE REVIEW

File: app/services/gov_uk_search_api/client.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    SEARCH_API_URL = 'https://www.gov.uk/api/search.json?'
+
+    def initialize(query)
+      @raw_query = query
+      @query = URI.encode(query)
+    end
+
+    def search
+      result = Curl.get('https://www.gov.uk/api/search.json?q=' + @query)
```
Comment (misaka) [inline app/services/gov_uk_search_api/client.rb]: Should this be `SEARCH_API_URL` here?

Comment (misaka) [inline app/services/gov_uk_search_api/client.rb]: Also, I can't help but wonder if that shouldn't be in the `settings.yml` file.

File: app/services/gov_uk_search_api/response.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    attr_reader :query, :result_items, :response_code
+
+    def initialize(query, raw_response)
+      @query = query
+      @raw_response = raw_response
+      @response_code = raw_response.response_code
+      @result_items = []
+      JSON.parse(raw_response.body_str)['results'].each do |result|
+        @result_items << ResultItem.new(result)
+      end
```
Comment (misaka) [inline app/services/gov_uk_search_api/response.rb]: This could be better expressed as a map:

```ruby
@result_items = JSON.parse(raw_response.body_str)['results'].map do |result|
  ResultItem.new(result)
end
```

File: app/services/gov_uk_search_api/client.rb (hunk 1 above)
Comment (stephenrichards) [inline app/services/gov_uk_search_api/client.rb]: Yes, it should be SEARCH_API_URL.  I don't think it should be in the settings file because it's fixed, and only used by this class.

File: app/services/gov_uk_search_api/response.rb (hunk 2 above)
Comment (stephenrichards) [inline app/services/gov_uk_search_api/response.rb]: cool

File: app/services/gov_uk_search_api/client.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    SEARCH_API_URL = 'https://www.gov.uk/api/search.json'
+
+    def initialize(query)
+      @raw_query = query
+      @query = URI.encode(query)
+    end
+
+    def search
+      result = Curl.get(SEARCH_API_URL + '?q=' + @query)
+      Response.new(@raw_query, result)
```
Comment (EdwardAndress) [inline app/services/gov_uk_search_api/client.rb]: Is there an advantage to breaking this down into two steps (initialize and search)?

I.e. Why not make search happen automatically after initialization, given the fact it's the only thing that can be done with an instance of Client?  Or how about making `#search` a class method?

File: spec/services/gov_uk_search_api/client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  RSpec.describe Client, type: :model do
+    describe '#search' do
+      it 'calls Curl with the URI encoded query' do
+        url_query = 'https://www.gov.uk/api/search.json?q=Ministry%20of%20Justice'
+        curl_easy = double Curl::Easy
+        response = double Response
+        expect(Curl).to receive(:get).with(url_query).and_return(curl_easy)
+        expect(Response).to receive(:new).with('Ministry of Justice', curl_easy).and_return(response)
+        actual_response = Client.new('Ministry of Justice').search
+        expect(actual_response).to eq response
```
Comment (EdwardAndress) [inline spec/services/gov_uk_search_api/client_spec.rb]: Could this last expectation ever fail?  The expectation and the stubbed response are the same... aren't they?

I.e. `actual_response` will be set to `response` because `Response.new` is stubbed to return `response`

File: spec/services/gov_uk_search_api/result_item_spec.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+      context 'display type of news story' do
+        it 'returns News Story with publication date and organisation acronyms' do
+          item = ResultItem.new(news_story_item)
+          expect(item.meta_data).to eq 'News story 3 Feb 2017 MOJ, HMCTS'
+        end
+      end
+    end
+
+    def standard_item
```
Comment (EdwardAndress) [inline spec/services/gov_uk_search_api/result_item_spec.rb]: Why are these methods rather than let blocks?

File: spec/services/gov_uk_search_api/client_spec.rb (hunk 4 above)
Comment (EdwardAndress) [inline spec/services/gov_uk_search_api/client_spec.rb]: Maybe I am too sleepy but I also find this a little hard to follow.  I think it's partly because the URI encoded query is referenced in the description but there's no clear reference to it in the block, so I have to work out which bit is the URI encoded query.  Is it the `url_query`?

Comment (stephenrichards) [inline spec/services/gov_uk_search_api/client_spec.rb]: The line that's under test is:
```Client.new('Ministry of Justice').search```
And this test is checking that a call to curl with the expected query string is made, and that the response from curl is used to instantiate a Response object, and it is that response object which is returned from the line under test,

File: spec/services/gov_uk_search_api/result_item_spec.rb (hunk 5 above)
Comment (stephenrichards) [inline spec/services/gov_uk_search_api/result_item_spec.rb]: No reason, but I tend to think let blocks are great for one liners, they become a bit hard to read when used to generate multi-line hashes as here.

File: spec/services/gov_uk_search_api/client_spec.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+require File.dirname(__FILE__) + '/../../rails_helper'
+
+module GovUkSearchApi
```
Comment (misaka) [inline spec/services/gov_uk_search_api/client_spec.rb]: I don't think I've ever seen `RSpec.describe` blocks being embedded into a module like this. Is there a benefit to doing it this way?

File: spec/services/gov_uk_search_api/client_spec.rb (hunk 4 above)
Comment (misaka) [inline spec/services/gov_uk_search_api/client_spec.rb]: I like how this test is quite isolated from other code with the use of doubles, etc. I also got tripped up by what got @EdwardAndress. What if we gave 'Ministry of Justice' meaning by assigning it a name:

```ruby
  query = 'Ministry of Justice'
  url_query = "https://www.gov.uk/api/search.json?q=#{URI.encode(query)}"
  ...
  expect(Response).to receive(:new).with(query, curl_easy).and_return(response)
  actual_response = Client.new(query).search 
```

File: Gemfile [hunk 7]
```diff
@@ -50,6 +51,7 @@ group :development, :test do
   gem 'rspec-rails', '~> 3.4'
   gem 'rubocop', require: false
   gem 'rubocop-rspec', require: false
+  gem 'awesome_print', '~> 1.7'
```
Comment (misaka) [inline Gemfile]: `awesome_print` looks ... awesome :) This list of gems is in alphabetical order, mind taking a moment to put this at the top?

File: app/services/gov_uk_search_api/response.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  class Response
+
+    attr_reader :query, :result_items, :response_code
+
+    def initialize(query, curl_response)
+      @query = query
+      @curl_response = curl_response
+      @response_code = @curl_response.response_code
+      @result_items = JSON.parse(@curl_response.body_str)['results'].map do |result|
+        ResultItem.new(result)
```
Comment (EdwardAndress) [inline app/services/gov_uk_search_api/response.rb]: Minor point: Is there a way to make the variable names consistent here too?

`result` and, later, `item_hash`
